### ISSUE TITLE
Redis in Cluster Mode, Getting SocketClosed message when Master goes down

### ISSUE THREAD

**Issue by murugappanm** [StackExchange/StackExchange.Redis#2409]
Hello,
We are running Redis in cluster mode with 3 Masters and 3 Slaves. We have a long running client that keeps reading/writing to the Redis through the StackExchange.Redis library (2.6.70)
When we tested with taking 1 master down, we noticed that the corresponding slave got promoted to being the Master in about a few seconds (confirmed using the redis cli). But, the long running client got no response from the SE.Redis and then the SE.Redis call from my client failed with a timeout. 
Does the StackExchange.Redis library provide some sort of retry mechanism with some wait if the current Master goes down? 
Are there any exceptions we can catch from the client side so that we can retry? Any help is appreciated.

**Comment by NickCraver** [comment]
> But, the long running client got no response from the SE.Redis and then the SE.Redis call from my client failed with a timeout.

Yep...that _is_ our retry mechanism most likely, commands go into a backlog to try sending if we can re-establish a connection within your configured timeout period.

You can catch these timeout exceptions and look at the [CommandStatus](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/Exceptions.cs#L47) to see if it's been sent to Redis yet (or timed out before it was ever sent), then decide to retry or not if you desire.

Side note: I have no idea what your scenario/desire is, e.g. is this for testing a failure case or for intentional maintenance events, but if you're changing nodes intentionally and have some control over it (e.g. normal patching), you can proactively inform all clients ASAP via pub/sub. You can do this with any `ConnectionMultiplexer` connected to that cluster, [calling `.PublishReconfigure(Async)`](https://github.com/StackExchange/StackExchange.Redis/blob/9698aaa8c5e372a73e0dec0adeeb9c85c09dd7f9/src/StackExchange.Redis/ConnectionMultiplexer.cs#L2127) which broadcasts to all SE.Redis clients to re-check their config, get a new `CLUSTER NODES` topology, etc. - all to adjust to changes as fast as possible.

**Comment by murugappanm** [comment]
Thanks Nick for your response.  We were running some chaos testing to understand our resiliency and that's when we ran into these timeouts. Right now, we are able to catch the RedisTimeoutException (and also the RedisServerException, and RedisConnectionException) and then retry.
I shall post an update after I rerun some tests to check the command status.

**Comment by NickCraver** [comment]
@murugappanm Any luck on your testing here? Curious how it played out.

**Comment by murugappanm** [comment]
@NickCraver Thanks for checking back. The command status was not sent when I checked. (WaitingToBeSent, if I remember that right)
Here is the scenario. I increased our client timeout to about 15 seconds and then reduced the cluster-node-timeout to 5 seconds and then started calling IO operations in a loop. Now when I stopped the master node manually in the linux server, the slave got promoted to master in about 5-6 seconds and the I/O calls still failed with timeout after 15 seconds. 

So, we decided to just catch the timeout exception (and also the RedisServerException, and RedisConnectionException) and then retry for now.
